### PR TITLE
Internal Kafka clients should do TLS hostname verification except Nodeports

### PR DESCRIPTION
**Pull request by see-quick**
Signed-off-by: see-quick <maros.orsak159@gmail.com>

### Type of change

- Bugfix

### Description

Currently, our internal Kafka clients do not perform TLS hostname verification on all possible Kubernetes listeners. This PR  solves https://github.com/strimzi/strimzi-kafka-operator/issues/7518 and enables TLS hostname verification and also disables it in situations where we do not want to execute it (e.g., when we use Nodeport listener, because it does not support TLS hostname verification on ports).  

### Checklist

- [x] Make sure all tests pass

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java [hunk 1]
```diff
@@ -155,6 +155,9 @@ private void testWeirdUsername(ExtensionContext extensionContext, String weirdUs
             .withMessageCount(MESSAGE_COUNT)
             .withPort(Constants.HTTP_BRIDGE_DEFAULT_PORT)
             .withNamespaceName(namespace)
+            // we disable ssl.endpoint.identification.algorithm for external listener (i.e., Nodeport),
+            // because TLS hostname verification is not supported on such listener type.
+            .withAdditionalConfig("ssl.endpoint.identification.algorithm=\n")
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:160]: Why does this use Node port listener here?

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java [hunk 2]
```diff
@@ -213,6 +216,9 @@ private void testWeirdUsername(ExtensionContext extensionContext, String weirdUs
             .withTopicName(topicName)
             .withMessageCount(MESSAGE_COUNT)
             .withUserName(weirdUserName)
+            // we disable ssl.endpoint.identification.algorithm for external listener (i.e., Nodeport),
+            // because TLS hostname verification is not supported on such listener type.
+            .withAdditionalConfig("ssl.endpoint.identification.algorithm=\n")
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: Same here ... Why does this use Node port listener here?

Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: The producer fails on this error:
```java
eadyOps=0] SSL Handshake failed
javax.net.ssl.SSLHandshakeException: No subject alternative names matching IP address 192.168.0.226 found
	at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:353) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:296) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:291) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1357) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1232) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1175) ~[?:?]
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392) ~[?:?]
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1074) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1061) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1008) ~[?:?]
```

Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: I'm not wondering about the disabling of it. I'm wondering why does this use node-ports in the first place as it seems unnecessary.

Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: I am not sure...you mean we should rather produce messages to the `internal listener` instead of producing them via external NodePort? 

Because these test cases currently do the following:
1. Create a bridge consumer to receive messages from the external listener (NodePort) 
2. Create a producer to produce messages via the same listener (NodePort)

So should we modify the second point to use an internal listener instead of an external one?

Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: Yeah, I think we should think about what the purpose of the test is. i mean I can approve this and we can merge it. But we should look into that in some separate PR - what exactly does this bring us.

Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: IIRC reason why we are using external producer etc. was because at the time I was writing those "weird username tests", there were no possibilities to use internal clients - as they didn't allow any username to be longer than some number of characters and it shouldn't contain any special characters.

External clients are allowing it. 

TBH I don't remember the exact reason why we went this way, but maybe we can rewrite those test to use our test-clients.

But the rewriting itself should be for a different PR.

**Comment by see-quick** [comment]
btw it seems that a few test cases are failing in ListenersST due to the TLS hostname verification...with following errror:
```java
Caused by: javax.net.ssl.SSLHandshakeException: No subject alternative DNS name matching my-cluster-82bc9931-kafka-bootstrap found.
	at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:353) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:296) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:291) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1357) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1232) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1175) ~[?:?]
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392) ~[?:?]
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1074) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1061) ~[?:?]
	at java.security.AccessC
```

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java (hunk 2 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:221]: > Yeah, I think we should think about what the purpose of the test is.

I agree, we definitely could do an investigation into such test cases and find out if they are necessary. :) I can create an issue for that.

**Comment by see-quick** [comment]
Update: I have managed to solve issue with the latest commit :)

**Review by Frawless — APPROVED** [review]

**Review by im-konge — APPROVED** [review]

File: systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java [hunk 3]
```diff
@@ -100,12 +100,29 @@ public static SystemTestCertAndKey generateIntermediateCaCertAndKey(SystemTestCe
...
                 .build();
     }
+
+    public static SystemTestCertAndKey generateIntermediateCaCertAndKey(final SystemTestCertAndKey rootCert,
+                                                                        final ASN1Encodable[] sanDnsNames) {
+        return intermediateCaCertBuilder(rootCert)
+                .withSubjectDn(STRIMZI_INTERMEDIATE_CA)
+                .withSanDnsNames(sanDnsNames)
+                .build();
+    }
```
Comment (scholzj) [inline systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java:110]: Why is this needed? Is this even used anywhere? The SANs in the CA should not matter.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java [hunk 4]
```diff
@@ -213,6 +213,9 @@ private void testWeirdUsername(ExtensionContext extensionContext, String weirdUs
             .withTopicName(topicName)
             .withMessageCount(MESSAGE_COUNT)
             .withUserName(weirdUserName)
+            // we disable ssl.endpoint.identification.algorithm for external listener (i.e., NodePort),
+            // because TLS hostname verification is not supported on such listener type.
+            .withAdditionalConfig("ssl.endpoint.identification.algorithm=\n")
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:218]: Why is this using node ports?

Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:218]: I think that we can re-write it to **not** use NodePort but such change I can be in another PR. The reason why it is using NodePort is answered by Lukas.

> image

I can definitely, open another PR, which will change it :)

File: systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java (hunk 3 above)
Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java:110]: We use it to generate:
1. Root CA
2. Intermediate CA
3. Strimzi CA

The intermediate CA is used for chain cert e.g.:
```java
SystemTestCertAndKey root1 = generateRootCaCertAndKey();
SystemTestCertAndKey intermediate1 = generateIntermediateCaCertAndKey(root1);
SystemTestCertAndKey strimzi1 = generateEndEntityCertAndKey(intermediate1);
ROOT_CA_CERT_AND_KEY_1 = exportToPemFiles(root1);
STRIMZI_CERT_AND_KEY_1 = exportToPemFiles(strimzi1);
CHAIN_CERT_AND_KEY_1 = exportToPemFiles(strimzi1, intermediate1, root1);
```

Comment (scholzj) [reply systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java:110]: The point is not whether you use Intermediate CA. The point is that you should not care about SANs for Intermediate CA (neither for Root CA). And you do not seem to use anywhere this particular method with the second parameter.

Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java:110]: Well, we use it in the test cases where I create whole chain cert (root, inter, strimzi) but I think as you mention the SANs should not matter so I will remove it.

Comment (see-quick) [reply systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertManager.java]: > The point is not whether you use Intermediate CA. The point is that you should not care about SANs for Intermediate CA (neither for Root CA). And you do not seem to use anywhere this particular method with the second parameter.

Yes, that's true and thanks for checking that :) <3

**Review by scholzj — APPROVED** [review]

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java (hunk 4 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:218]: I have created issue to have it tracked -> https://github.com/strimzi/strimzi-kafka-operator/issues/7608

**Comment by see-quick** [comment]
I noticed that the test case `testCaRenewalBreakInMiddle` is failing for some reason. (but it is not related to my PR). I can create issue for fixing that test case :).
